clang/lib/Serialization/ASTReader.cpp

```diff
@@ -7392,27 +7388,28 @@ QualType ASTReader::GetType(TypeID ID) {
   return TypesLoaded[Index].withFastQualifiers(FastQuals);
 }
 
-QualType ASTReader::getLocalType(ModuleFile &F, unsigned LocalID) {
+QualType ASTReader::getLocalType(ModuleFile &F, TypeID LocalID) {
```

Comment:
Could we clarify what `LocalID` means here in a comment somewhere?
IIUC, the `ModuleFileIndex` there is an index into `F.TransitiveImports` rather than into a "global" module manager.

Previously, we had distinction between `unsigned` in parameters to this function and `TypeID` as a return type.
Maybe we should keep this here and accept `uint64_t` to avoid confusing the common global module file index and a local file index?

Reply:
> Could we clarify what LocalID means here in a comment somewhere?

In ASTReader.h, there is a single line comment: `Resolve a local type ID within a given AST file into a type.`. I add a new line to explain it as "A local type ID is only meaningful with the corresponding module file. See the implementation of getGlobalTypeID for details.".

> IIUC, the ModuleFileIndex there is an index into F.TransitiveImports rather than into a "global" module manager.

Yes, this is correct for local ID. But I feel it might be odd to put this in the comment. I feel, it might be good to know, we can use `ASTReader::getGlobalTypeID` to convert a `<ModuleFile, LocalTypeID>` pair to a `GlobalTypeID`. And the concrete process is the implementation details to GlobalTypeID`.

> Previously, we had distinction between unsigned in parameters to this function and TypeID as a return type.
Maybe we should keep this here and accept uint64_t to avoid confusing the common global module file index and a local file index?

I don't feel it better. `uint64_t` doesn't provide any type information. To be best, maybe we can do something like for `LocalDeclID` and `GlobalDeclID`. I only did that for DeclID since I feel the DeclID is the most complicated and the TypeID are much more simpler. I think we can this later if we feel it is necessary.

Reply:
The reason I feel `TypeID` is the "wrong" type here is because after reading the comments it feels that the module file index is always global. Keeping `TypeID` here is probably okay, but we should put this distinction into a comment somewhere (either a comment to `getLocalType` that says the TypeID we pass here is special because the index is relative to `F` or a comment to the `TypeID` itself that explains that module file index can be either global or relative to a module file).
I would probably vouch for putting this into `getLocalType` to avoid overloading the readers of the code that does not touch `getLocalType` with this special case (it seems to be quite limited anyway, we quickly read the value from the AST and convert it right away, same for writing)

Reply:
Agreed. I tried to address this by introducing a new type alias `using LocalTypeID = TypeID;` with a FIXME that we'd better to make them classes like we did for DeclID. Then at least I believe the interface will have better readability.